cadcdata/cadcdata/core.py

```diff
@@ -0,0 +1,418 @@
+        :param wcs: True if the wcs is to be included with the file
+        :param process_bytes: function to be applied to the received bytes
+        :return: the data stream object
+        """
+        assert archive is not None
+        assert file_id is not None
+        resource = '/{}/{}'.format(archive, file_id)
+        params = {}
+        if fhead:
+            params['fhead'] = fhead
+        if wcs:
+            params['wcs'] = wcs
+        if cutout:
+            params['cutout'] = cutout
+        self.logger.debug('GET '.format(resource))
+        response = self._repo_client.get(resource, stream=True, params=params)
+        if file is not None:
+            if not hasattr(file, 'read'):
+                # got a file name?
+                with open(file, 'w') as f:
+                    self._save_bytes(response, f, decompress, process_bytes)
+            else:
+                self._save_bytes(response, file, decompress, process_bytes)
+        else:
+            # get the file name from the content disposition
+            content_disp = response.headers.get('content-disposition', '')
+            file = file_id
+            for content in content_disp.split():
+                if 'filename=' in content:
+                    file = content[9:]
+            if file.endswith('gz') and decompress:
+                file = os.path.splitext(file)[0]
+            self.logger.debug('Using content disposition file name: {}'.format(file))
+            with open(file, 'w') as f:
+                self._save_bytes(response, f, decompress, process_bytes)
+        self.logger.debug('Successfully saved file\n')
+
+
+    def _save_bytes(self, response, file, decompress=False, process_bytes=None):
+        # requests automatically decompresses the data. Tell it to do it only if it had to
+        if not decompress:
+            chunk = response.raw.read(READ_BLOCK_SIZE)
+            while len(chunk) > 0:
+                if process_bytes is not None:
+                    process_bytes(chunk)
+                file.write(chunk)
+                chunk = response.raw.read(READ_BLOCK_SIZE)
+        else:
+            chunk = response.iter_content(READ_BLOCK_SIZE)
+            while len(chunk) > 0:
+                if process_bytes is not None:
+                    process_bytes(chunk)
+                file.write(chunk)
+                chunk = response.iter_content(READ_BLOCK_SIZE)
+
+
+    def put_file(self, archive, file_id, file, archive_stream=None):
+        """
+        Puts a file into the archive storage
+        :param archive: name of the archive
+        :param file_id: ID of file in the archive storage
+        :param file: location of the source file
+        :param archive_stream: specific archive stream
+        """
+        assert archive is not None
+        assert file_id is not None
+        resource = '/{}/{}'.format(archive, file_id)
+        self.logger.debug('PUT {}'.format(resource))
```

Comment:
Doesn't the data PUT service have more parameters than this?   I thought you could set the file_id or leave blank and then service would determine based on the filename?